### ISSUE TITLE
iOS Memory Leak when playing OGG / WAV files

### ISSUE THREAD

**Issue by jskuse** [haxenme/nme#446]
I've been running an updated version of our game on the latest GIT NME through Instruments on iOS and noticed that memory use was maybe running higher than usual. I've narrowed this down to a seemingly significant leak of OpenAL sounds. Or maybe this isn't really new and I've just never noticed before - it's unclear.

If you take the following (intentionally very inefficient) code:
```
class Main extends nme.display.Sprite {
	private var _count:Int = 0;

	public function new () {
		super();
		stage.color = 0x00FF00;
		stage.addEventListener(nme.events.Event.ENTER_FRAME, enterFrame);
	}

	private function enterFrame(e:nme.events.Event):Void {
		if (_count++ % 3 == 0) {
			var sound:nme.media.Sound = nme.Assets.getSound("assets/test.ogg", false);
			if (sound != null) { sound.play(); }
		}
	}
}
```

...and profile it, you get a memory profile like this (the graph terminates when it eventually crashes):
![graph](https://cloud.githubusercontent.com/assets/6178279/22657862/e425cd64-ec8f-11e6-8612-ec71a372e23e.png)

With the primary culprits being:
![summary](https://cloud.githubusercontent.com/assets/6178279/22657877/e99bc820-ec8f-11e6-9759-2f255cdeeaab.png)

All of which seem to track back to OpenAL:
![details](https://cloud.githubusercontent.com/assets/6178279/22657881/ecb69a4e-ec8f-11e6-95e5-9022179696dc.png)

I did a bit of printf style debugging and I'm not seeing the ```OpenALBufferChannel```'s destructor ever getting called, even after the ```ChannelList``` appears to remove the channels. What should be responsible for cleaning up the channel objects? Is it some HXCPP GC magic that I don't understand or is it explicit somewhere on the NME C++ side that I haven't spotted?

**Comment by jskuse** [comment]
I've had another poke around this afternoon on this issue. Having had a look at some typical clear-up code for OpenAL, I was able to fix the 16kb leak by explicitly unbinding the buffer from the source and then deleting the source in ```OpenALSourceChannel``` e.g.  

```
   void stop()
   {
      clRemoveChannel(this);
      if (!openal_is_shutdown && sourceId)
      {
         if (playing()) 
            alSourceStop(sourceId);

         alSourcei(sourceId, AL_BUFFER, 0);
         alDeleteSources(1, &sourceId);
      }
      sourceId = 0;

      if (soundObject)
      {
         soundObject->DecRef();
         soundObject = 0;
      }
   }
   ```

No idea if that's right, but it seems to work. The larger allocations seem to be from the buffer itself never being freed because ```OpenALSound```'s destructor doesn't *appear* to get called. I tested this hypothesis by hacking in some code to free the buffer elsewhere and observed that the memory build up was eliminated, but still can't for the life of me get my head around what is supposed to be responsible for deleting the Sound objects.

**Comment by madrazo** [comment]
I believe that the OpenAL object is deleted only when the haxe Sound is removed by the Garbage Collector. Maybe you can add a  cpp.vm.GC.run( true ) to check it.

**Comment by jskuse** [comment]
Thanks @madrazo - I've already tried forcing the GC in various ways, but no luck, it just doesn't *seem* to be running the Sound's destructor and so the buffer isn't getting freed (it seems to me that the Sound owns the AL buffer whereas Channel owns the AL source). I suppose the issue here is that I don't get how to debug this because I don't understand how the C++ object's lifecycle is controlled (reference counting? black magic?) but maybe I just need to look at the code some more!

**Comment by jskuse** [comment]
I've submitted a pull request for this (#447) now.


If anyone accidentally ends up here from a search engine wondering the same as I did about the lifecycle of the NME C++ objects, my comb through the code suggests that it appears to be a combination of both reference counting *and* magic (disclaimer: I could be wrong about any of this):

A core ```Object``` type (defined in Object.h) deals with reference counting, proving ```IncRef``` and ```DecRef``` functions which are called on the C++ side as required. So a SoundChannel will ```IncRef``` the underlying Sound when it starts playing and ```DecRef``` when it finishes, preventing the Sound going anywhere while it's being used by the SoundChannel. ```DecRef```ing to zero is ultimately the thing that calls ```delete``` on the Object.

The magicish bit comes via the various functions in ExternalInterface.cpp which interfaces with the Haxe side of things, which runs Objects through the ```ObjectToAbstract``` function (defined in NmeCffi.h) before passing them out. This uses the ```val_gc``` function to register a function called ```release_object``` with the GC. This function is called when the GC decides it's done with the Object, which then ```DecRef```s it, potentially resulting in a ```delete``` if the reference count has dropped to zero.
